File: codeforces-1514-d-INCOMPLETE.java
import java.io.*;
import java.util.*;

class Main {
    private int n;
    private int[] arr;
    private Map<Integer, List<Integer>> map;
    private Random rand;

    public Main(int n, int[] arr, Map<Integer, List<Integer>> map) {
        this.n = n;
        this.arr = arr;
        this.map = map;
        this.rand = new Random();
    }

    public int bin(int pick, int val) {
        List<Integer> num_freq = this.map.get(pick);

        int l = -1; int r = num_freq.size();
        while(r > l + 1) {
            int m = (l + r) / 2;
            if(num_freq.get(m) <= val) {
                l = m;
            } else {
                r = m;
            }
        }

        return l;
    }

    public int cutAndStick(int l, int r) {
        int max_freq = (r - l + 2) / 2;
        int[] picks = new int[25];
        
        for(int i=0; i<25; i++) {
            int tmp = this.rand.nextInt(r-l+1) + l;
            picks[i] = this.arr[tmp];
        }

        for(int i=0; i<25; i++) {
            int lower = this.bin(picks[i], l-1);
            int upper = this.bin(picks[i], r);
            if(upper - lower > max_freq) {
                return 2 * (upper - lower) - r + l - 1;
            }
        }

        return 1;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken()); int q = Integer.parseInt(st.nextToken());
        int[] arr = new int[n];
        Map<Integer, List<Integer>> map = new HashMap<Integer, List<Integer>>();
        st = new StringTokenizer(br.readLine());
        for(int i=0; i<n; i++) {
            arr[i] = Integer.parseInt(st.nextToken());
            map.putIfAbsent(arr[i], new ArrayList<Integer>());
            map.get(arr[i]).add(i);
        }

        Main sol = new Main(n, arr, map);
        StringBuilder ans = new StringBuilder();
        for(int z=0; z<q; z++) {
            st = new StringTokenizer(br.readLine());
            int l = Integer.parseInt(st.nextToken()); int r = Integer.parseInt(st.nextToken());
            l--; r--;
            ans.append(sol.cutAndStick(l, r));
            ans.append("\n");
        }
        System.out.print(ans.toString());
    }
}

File: codeforces-1514-d-INCOMPLETE.py
import sys
import random
import io, os

input = io.BytesIO(os.read(0, os.fstat(0).st_size)).readline

class Soln():
    def __init__(self, n, arr):
        self.n = n
        self.arr = arr
        self.d = {}
        for i in range(n):
            self.d.setdefault(arr[i], list()).append(i)

    def bin(self, pick, val):
        num_freq = self.d[pick]
        
        l, r = -1, len(num_freq)
        while r > l + 1:
            m = (l + r) // 2
            if num_freq[m] <= val:
                l = m
            else:
                r = m

        return l

    def cutAndStick(self, l, r):
        max_freq = (r - l + 2) // 2
        picks = []
        for _ in range(25):
            i = random.randint(l, r)
            picks.append(self.arr[i])

        for p in picks:
            lower = self.bin(p, l-1)
            upper = self.bin(p, r)
            if upper - lower > max_freq:
                return str(2 * (upper - lower) - (r - l + 1))

        return '1'


def main():
    n, q = map(int, input().split())
    arr = [int(i) for i in input().split()]
    s = Soln(n, arr)
    ans = []
    for _ in range(q):
        l, r = map(int, input().split())
        l -= 1; r -= 1
        ans.append(s.cutAndStick(l, r))
    print('\n'.join(ans))


if __name__ == '__main__':
    main()